handlers/mapping_handlers.go

```diff
@@ -0,0 +1,404 @@
+	}
+
+	vars := mux.Vars(r)
+	conversationID, err := strconv.ParseInt(vars["conversation_id"], 10, 64)
+	if err != nil {
+		errMsg := "Invalid conversation ID"
+		log.Println(errMsg + ": " + err.Error())
+		http.Error(w, errMsg, http.StatusBadRequest)
+		return
+	}
+
+	sessionMember, err := env.DB.GetUserConversationMapping(userID, conversationID)
+	if err != nil {
+		internalServerError(w, err)
+		return
+	}
+	if sessionMember == nil {
+		errMsg := fmt.Sprintf("User %d is not in conversation %d", userID, conversationID)
+		log.Println(errMsg)
+		http.Error(w, "Conversation not found", http.StatusNotFound)
+		return
+	}
+
+	members, err := env.DB.GetUserConversationMappings(conversationID)
+	if err != nil {
+		internalServerError(w, err)
+		return
+	}
+	memberList := &models.UserConversationMappingList{Users: members}
+
+	// TODO: Get user names from karen?
+
+	w.Header().Add("Content-Type", "application/json")
+	json.NewEncoder(w).Encode(memberList)
+}
+
+// PatchMappingHandler updates a single user in a conversation
+func (env *Env) PatchMappingHandler(w http.ResponseWriter, r *http.Request) {
+	defer r.Body.Close()
+	userID, err := strconv.ParseInt(r.Header.Get("User-ID"), 10, 64)
+	if err != nil {
+		errMsg := "Invalid user ID"
+		log.Println(errMsg + ": " + err.Error())
+		http.Error(w, errMsg, http.StatusBadRequest)
+		return
+	}
+
+	vars := mux.Vars(r)
+	conversationID, err := strconv.ParseInt(vars["conversation_id"], 10, 64)
+	if err != nil {
+		errMsg := "Invalid conversation ID"
+		log.Println(errMsg + ": " + err.Error())
+		http.Error(w, errMsg, http.StatusBadRequest)
+		return
+	}
+	memberID, err := strconv.ParseInt(vars["user_id"], 10, 64)
+	if err != nil {
+		errMsg := "Invalid user ID"
+		log.Println(errMsg + ": " + err.Error())
+		http.Error(w, errMsg, http.StatusBadRequest)
+		return
+	}
+	if userID == memberID {
+		errMsg := "Cannot modify own role"
+		log.Println(errMsg)
+		http.Error(w, errMsg, http.StatusForbidden)
+		return
+	}
+
+	conversation, err := env.getConversation(w, conversationID)
+	if err != nil || conversation == nil {
+		return
+	}
+
+	sessionMember, err := env.DB.GetUserConversationMapping(userID, conversationID)
+	if err != nil {
+		internalServerError(w, err)
+		return
+	}
+	if sessionMember == nil {
+		errMsg := fmt.Sprintf("User %d is not in conversation %d", userID, conversationID)
+		log.Println(errMsg)
+		http.Error(w, "Conversation not found", http.StatusNotFound)
+		return
+	}
+
+	member, err := env.DB.GetUserConversationMapping(memberID, conversationID)
+	if err != nil {
+		internalServerError(w, err)
+		return
+	}
+	if member == nil {
+		errMsg := fmt.Sprintf("User %d is not in conversation %d", userID, conversationID)
+		log.Println(errMsg)
+		http.Error(w, "User not found", http.StatusNotFound)
+		return
+	}
+
+	reqMember := &models.UserConversationMapping{}
+	if err := parseMappingJSON(w, r.Body, reqMember); err != nil {
+		return
+	}
+
+	if reqMember.Role == "" && reqMember.Nickname == nil {
```

Comment:
I thought that these 2 fields would get updated based on receiving messages on a queue, rather than from a client request. But, I guess maybe it does make more sense for clients to update these things themselves.

Reply:
Hmm. That's a good point. I think `LastOpened` could be handled by a message queue so when a client closes a WebSocket, i.e. closes a conversation, `patches` could put a message on the queue to be like "this user closed the conversation" so `ether` would update it. `Pending` I think can be purely client, since if they have some pending invitation, when they accept, the client can just send a PATCH to update it.